### PR TITLE
Verify packageName in bundle matches repository name

### PR DESCRIPTION
**Pull request by jeremylinlin**
Verification step added to ensure the packageName defined in package.yaml matches the quay.io repo name given as a parameter to the push command.

### CODE REVIEW

File: operatorcourier/api.py [hunk 1]
```diff
@@ -19,14 +19,16 @@
...
-def build_and_verify(source_dir=None, yamls=None, ui_validate_io=False):
+def build_and_verify(source_dir=None, yamls=None, ui_validate_io=False, repository=None):
     """Build and verify constructs an operator bundle from
     a set of files and then verifies it for usefulness and accuracy.
 
     It returns the bundle as a string.
 
     :param source_dir: Path to local directory of yaml files to be read.
     :param yamls: List of yaml strings to create bundle with
+    :param ui_validate_io: TODO
```
Comment (kevinrizza) [inline operatorcourier/api.py]: Let's leave documentation for this parameter for a separate pr please.

File: operatorcourier/validate.py [hunk 2]
```diff
@@ -208,43 +216,48 @@ def _pkgs_validation(self, bundleData):
...
-                        try:
-                            csvName = csv["metadata"]["name"]
-                            csvNames.append(csvName)
-                        except KeyError:
-                            pass
-                    for channel in channels:
-                        if "name" not in channel:
-                            logger.error("package channel.name not defined.")
-                            valid = False
+        pkg = pkgs[0]
```
Comment (kevinrizza) [inline operatorcourier/validate.py]: What about the case where pkgs == 0 ?

Comment (jeremylinlin) [reply operatorcourier/validate.py]: @kevinrizza It should already be handled [here](https://github.com/operator-framework/operator-courier/blob/master/operatorcourier/validate.py#L254) before the pkg validator is called.

File: operatorcourier/api.py (hunk 1 above)
Comment (jeremylinlin) [reply operatorcourier/api.py]: Sure. Am I supposed to remove this line?

Comment (kevinrizza) [reply operatorcourier/api.py]: that is what I mean, yes.

File: operatorcourier/validate.py (hunk 2 above)
Comment (kevinrizza) [reply operatorcourier/validate.py]: @jeremylinlin we can't rely on other validation methods. what if they were reordered or that other method was refactored?

Comment (jeremylinlin) [reply operatorcourier/validate.py]: @kevinrizza Ah that makes sense. I'll fix it. Thanks Kevin!

File: operatorcourier/validate.py [hunk 3]
```diff
@@ -232,43 +240,49 @@ def _pkgs_validation(self, bundleData):
...
 
-        for pkg in pkgs:
-            if "packageName" in pkg:
-                logger.info("Evaluating package %s", pkg["packageName"])
-            else:
-                self._log_error("packageName not defined.")
-                valid = False
+        num_pkgs = len(pkgs)
+        if num_pkgs != 1:
+            logger.error('Only 1 package is expected to exist per bundle, but got %d.', num_pkgs)
```
Comment (awgreene) [inline operatorcourier/validate.py]: This should use the _log_error method

Comment (jeremylinlin) [reply operatorcourier/validate.py]: Thanks @awgreene. Great catch! Updated the PR.

File: operatorcourier/validate.py [hunk 4]
```diff
@@ -75,6 +76,14 @@ def _bundle_validation(self, bundle):
...
             validationDict[self.csvKey] &= self._type_validation(
                 bundleData, self.csvKey, self._ui_validation_io, True)
+        if validationDict[self.pkgsKey] and repository is not None:
+            packageName = bundleData['packages'][0]['packageName']
+            match = repository == packageName
+            if not match:
+                self._log_error('The packageName (%s) in bundle does not match '
+                                'repository name (%s) provided as command line argument.',
+                                packageName, repository)
+            validationDict[self.pkgsKey] &= match
```
Comment (kevinrizza) [inline operatorcourier/validate.py]: @jeremylinlin 
Why are we doing this bitwise and here? shouldn't we only define behavior in the case that they don't match?

Comment (jeremylinlin) [reply operatorcourier/validate.py]: Thanks @kevinrizza, it makes sense. I have simplified the code.

File: operatorcourier/validate.py [hunk 5]
```diff
@@ -45,18 +45,19 @@ def _log_error(self, message, *args, **kwargs):
...
         logger.error(message, *args, **kwargs)
 
-    def validate(self, bundle):
+    def validate(self, bundle, repository=None):
         """validate takes a bundle as a dictionary and returns a boolean value that
         describes if the bundle is valid. It also logs verification information when
         there are invalid sections in the bundle.
 
-        :param bundle: Dictionary of bundle value
+        :param bundle:     Dictionary of bundle value
```
Comment (SamiSousa) [inline operatorcourier/validate.py]: nit: Does this indent need to match for all params?

File: operatorcourier/validate.py [hunk 6]
```diff
@@ -75,6 +76,13 @@ def _bundle_validation(self, bundle):
         if self.ui_validate_io:
             validationDict[self.csvKey] &= self._type_validation(
                 bundleData, self.csvKey, self._ui_validation_io, True)
+        if validationDict[self.pkgsKey] and repository is not None:
```
Comment (SamiSousa) [inline operatorcourier/validate.py]: So this check will only run for the `push` command, correct?

**Comment by kevinrizza** [comment]
> One note: I know we don't want to introduce breaking changes, but is it possible to have a CLI flag that let's us skip entering the name operator bundle repo and instead uses the packageName found in the bundle?This is a nice-to-have, @kevinrizza let me know what you think.

@SamiSousa I don't want to do this because, while we technically have a soft requirement that packageName and repository name match, that's something that could be removed later. I would prefer not to introduce a feature that we may have to remove later.

File: operatorcourier/validate.py (hunk 5 above)
Comment (jeremylinlin) [reply operatorcourier/validate.py]: @SamiSousa It seems that we don't have this indent in other places. I'll fix it.

File: operatorcourier/validate.py (hunk 6 above)
Comment (jeremylinlin) [reply operatorcourier/validate.py]: That's correct. Because the repository argument is only required for push command.
